### PR TITLE
<DatePicker/> Split into separate internal components: <Calendar/> and <InputDate/>

### PR DESCRIPTION
**Pull request by kranux**
### What changed

...

### Why it changed

...

---

- [ ] Change is tested
- [ ] Change is documented
- [ ] Build is green
- [ ] Change of UX is approved by [wuwa](https://github.com/wuwa) or [milkyfruit](https://github.com/milkyfruit)

### CODE REVIEW

File: src/DatePicker/Calendar/Calendar.js [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    this.state = {
+      isMonthPickerOpen: false,
+      isYearPickerOpen: false,
+      month: props.value
+    };
+  }
+
+  // TODO: Change to getDerivedStateFromProps with React ^16.0.0
+  componentWillReceiveProps(nextProps) {
+    this.setState({month: nextProps.value});
```
Comment (Faradey27) [inline src/DatePicker/Calendar/Calendar.js]: this. _setMonth(nextProps.value)

File: src/DatePicker/Calendar/Calendar.js [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+  value: PropTypes.object,
+
+  /** Display a selectable yearDropdown */
+  showYearDropdown: PropTypes.bool,
+
+  /** Display a selectable monthDropdown */
+  showMonthDropdown: PropTypes.bool,
+
+  /** should the calendar close on day selection */
+  shouldCloseOnSelect: PropTypes.bool,
```
Comment (argshook) [inline src/DatePicker/Calendar/Calendar.js]: that should no longer be responsibility of `<Calendar/>`

File: src/DatePicker/DatePickerHead/styles.scss [hunk 3]
```diff
@@ -4,35 +4,38 @@
...
+  color: $B10;
 }
 
 .caption {
   padding: 0 6px;
-  color: $D80;
+  color: $D10;
   font-weight: 300;
-  @include FontLight();
+  @include FontRoman();
```
Comment (argshook) [inline src/DatePicker/DatePickerHead/styles.scss]: perhaps we should use `<Text/>`, if possible

File: src/DatePicker/Calendar/Calendar.js [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            />
+        }
+      </div>
+    );
+  }
+}
+
+Calendar.propTypes = {
+  /** Should show or hide the component */
+  visible: PropTypes.bool,
```
[the remark this answers is not in the document]
Comment (xmityaz) [reply src/DatePicker/Calendar/Calendar.js]: Added this prop for consistency with `DropdownLayout` component. The basic idea of these components is the same, so it looks logically to keep the same prop for visibility

File: src/DatePicker/Calendar/Calendar.js [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+      this._focusCalendarDay(this.props.value || new Date());
+    }
+  }
+
+  render() {
+    const {visible} = this.props;
+
+    return (
+      <div className={styles.root}>
+        {visible &&
```
[the remark this answers is not in the document]
Comment (xmityaz) [reply src/DatePicker/Calendar/Calendar.js]: the same reason as above
Do you think it is better to get rid of these props?

File: src/DatePicker/Calendar/Calendar.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (argshook) [reply src/DatePicker/Calendar/Calendar.js]: i don't think `<Calendar/>` needs to know if it's visible or not,
whatever the parent of calendar is should be responsible for that.

at some point `<Calendar/>` should be used as building block for other
calendar layouts (2month, in modal, for range etc.) and so those parents
will need to handle visibility anyway. moreower, the implementation just
becomes a tiny bit simpler :)

**Comment by milkyfruit** [comment]
Discussed with @kranux today about the focus tab-index between `<InputDate />` and `<Calendar />`.  I'll prepare the specs first, once it's ready then we can move to development.

**Comment by kranux** [comment]
Some info for future maintainers of this code. I noticed that we have an issue with calculating popover position / direction. For example:

image

After clicking on range input, calendar popover opens to bottom of input:

image

Even there seems to be enough room on top.

Seems that this issue is related with focus on input field which causes in multiple position recalculations in popper.js. To avoid this, tried adding focus to calendar element and listening to key event on window (which might cause event more issues, as we need to redirect all the events to calendar component).

**Comment by Faradey27** [comment]
Leftovers:
[UI](https://github.com/wix/wix-style-react/pull/2030#pullrequestreview-141411406)
- [ ] Show hover effect when today is focused
- [ ] Show keyboard focus effect only when keyboard is clicked. Don't reveal it on a mouse click

[Dev](https://github.com/wix/wix-style-react/pull/2030#discussion_r205784008)
- [ ] Calendar should not know about is it open or not
- [ ] [Popover position](https://github.com/wix/wix-style-react/pull/2030#issuecomment-408820489)

**Comment by Faradey27** [comment]
```html
<!DOCTYPE html>
<html>
<head>
   <title>DatePicker</title>
   <style type="text/css">

      body {
         background-color: #F0F4F7;
         text-align: center;
      }
      div {
         display: inline-block;
         margin-top: 78px;
         padding: 24px;
         background-color: white;
         border-radius: 8px;
         box-shadow: 0 6px 6px 0 rgba(22, 45, 61, 0.06), 0 0 18px 0 rgba(22, 45, 61, 0.12);
      }
      table {
         font-family: "Helvetica Neue";
         font-size: 14px;
         line-height: 18px;
         font-weight: 300;
         color: #162D3D;
         background-color: white;
         text-align: center;
         border-spacing: 0;
         margin: 0;
         padding: 0;
      }
      td {
         margin: 0;
         padding: 0;
         border-collapse: collapse;
      }
      h4 {
         font-family: "Helvetica Neue";
         font-size: 16px;
         line-height: 24px;
         font-weight: normal;
         color: #162D3D;
         margin: 0;
         padding: 6px 0 6px 0;
      }
      span {
         display: inline-block;
         padding: 18px 0 12px 0;
      }
      button {
         cursor: pointer;
         position: relative;
         z-index: 1;
         height: 36px;
         width: 36px;
         font-family: "Helvetica Neue";
         font-size: 14px;
         line-height: 18px;
         font-weight: 300;
         border: none;
         border-radius: 18px;
         transition: background 150ms;
      }
      button:hover {
         background-color: #EAF7FF;
      }
      button:focus {
         z-index: 2;
         outline: none;
         box-shadow: 0 0 0 3px #AADBFC;
      }
      button:active {
         box-shadow: none;
      }
      button.selected {
         color: white;
         background-color: #3899EC;
      }
      button.selected:hover {
         background-color: #4EB7F5;
      }
      button.selected:active {
         background-color: #3899EC;
         box-shadow: none;
      }
      button.today {
         background-color: #F0F4F7;
         box-shadow: inset 0 0 0 3px white;
      }
      button.today:hover {
         background-color: #EAF7FF;
         box-shadow: none;
      }
      button.today:focus {
         box-shadow: inset 0 0 0 3px white,
                     0 0 0 3px #AADBFC
         ;
      }
      button.today:hover:focus {
         background-color: #EAF7FF;
         box-shadow: 0 0 0 3px #AADBFC;
      }
      button.today:hover:active {
         background-color: #EAF7FF;
         box-shadow: none;
      }
      button.range-start {
         border-radius: 18px 0 0 18px;
         background-color: #EAF7FF;
      }
      button.range-end {
         border-radius: 0 18px 18px 0;
         background-color: #EAF7FF;
      }
      button.range {
         border-radius: 0;
         background-color: #EAF7FF;
      }

   </style>
</head>
<body>

   <div>
      <TABLE>
         <tr>
            <td colspan="7"><h4>February 2018</h4></td>
         </tr>
         <tr>
            <td><span>Su</span></td>
            <td><span>Mo</span></td>
            <td><span>Tu</span></td>
            <td><span>We</span></td>
            <td><span>Th</span></td>
            <td><span>Fr</span></td>
            <td><span>Sa</span></td>
         </tr>
         <tr>
            <td><button>1</button></td>
            <td><button>2</button></td>
            <td><button>3</button></td>
            <td><button>4</button></td>
            <td><button>5</button></td>
            <td><button>6</button></td>
            <td><button>7</button></td>
         </tr>
         <tr>
            <td><button>8</button></td>
            <td><button class="today">9</button></td>
            <td><button>10</button></td>
            <td><button>11</button></td>
            <td><button class="selected">12</button></td>
            <td><button>13</button></td>
            <td><button>14</button></td>
         </tr>
         <tr>
            <td><button>15</button></td>
            <td><button>16</button></td>
            <td><button>17</button></td>
            <td><button>18</button></td>
            <td><button>19</button></td>
            <td><button>20</button></td>
            <td><button>21</button></td>
         </tr>
         <tr>
            <td><button>22</button></td>
            <td><button>23</button></td>
            <td><button>24</button></td>
            <td><button>25</button></td>
            <td><button>26</button></td>
            <td><button>27</button></td><!-- 
            <td><button class="range-start">24</button></td>
            <td><button class="range">25</button></td>
            <td><button class="range">26</button></td>
            <td><button class="range-end">27</button></td> -->
            <td><button>28</button></td>
         </tr>
      </TABLE>
   </div>
   <script type="text/javascript">
      document.addEventListener('click', function(e) {
         if(document.activeElement.toString() == '[object HTMLButtonElement]'){
            document.activeElement.blur();
         }
      });
   </script>
</body>
</html>
```
awesome example about how things should be from @milkyfruit , related to UI leftovers

**Comment by tzabarc** [comment]
can't reproduce popover position issue. it seems to be at up position when it can.

**Pull request merged**